src/objectives/regularizers.jl

```diff
@@ -1,154 +1,175 @@
+            ∇[Δt_indices] .+= ∇Δt
         end
-        return structure
     end
+    
+    return nothing
+end
 
-    function ∂²L(Z⃗::AbstractVector) 
-        values = []
-        # Match Hessian structure indices
-        for t ∈ times
-            if traj.timestep isa Symbol
-                Δt = Z⃗[slice(t, traj.components[traj.timestep], traj.dim)]
-                append!(values, R .* Δt.^2)
-                # ∂²_vₖ_Δt, ∂²_Δt_vₖ
-                vₖ = Z⃗[slice(t, traj.components[name], traj.dim)]
-                Δv = vₖ .- baseline[:, t]
-                append!(values, 2 * (R .* (Δt .* Δv)))
-                append!(values, 2 * (R .* (Δt .* Δv)))
-                # ∂²_Δt_Δt
-                append!(values, Δv' * (R .* Δv))
-            else
-                Δt = traj.timestep
-                append!(values, R .* Δt.^2)
-            end
-        end
-        return values
+function hessian_structure(reg::QuadraticRegularizer, traj::NamedTrajectory)
+    Z_dim = traj.dim * traj.N + traj.global_dim
+    structure = spzeros(Z_dim, Z_dim)
+    
+    v_comps = traj.components[reg.name]
+    Δt_comp = traj.components[traj.timestep][1]
+    
+    for k ∈ reg.times 
+        v_indices = slice(k, v_comps, traj.dim)
+        Δt_index = index(k, Δt_comp, traj.dim)
+
+        # ∂²J/∂v²
+        structure[v_indices, v_indices] .= 1.0
+
+        # ∂²J/∂Δt∂v
+        structure[v_indices, Δt_index] .= 1.0
+
+        # ∂²J/∂Δt²
+        structure[Δt_index, Δt_index] = 1.0
     end
+    
+    return structure
+end
 
-    return Objective(L, ∇L, ∂²L, ∂²L_structure)
+function get_full_hessian(reg::QuadraticRegularizer, traj::NamedTrajectory)
+    Z_dim = traj.dim * traj.N + traj.global_dim
+    ∂²J = spzeros(Z_dim, Z_dim)
+    
+    v_comps = traj.components[reg.name]
+    Δt_comp = traj.components[traj.timestep][1]
+    
+    for t ∈ reg.times
+        zₖ = traj[t]
+        Δt = zₖ.timestep
+        v_indices = slice(t, v_comps, traj.dim)
+        Δt_index = index(t, Δt_comp, traj.dim)
+
+        rₖ = zₖ[reg.name] - reg.baseline[:, t]
+
+        # ∂²J/∂v² = Δt² * R
+        ∂²J[v_indices, v_indices] = Δt^2 * spdiagm(reg.R)
+
+        # ∂²J/∂Δt∂v = Δt * R * (v - baseline)
+        ∂²J[v_indices, Δt_index] = 2 * Δt * reg.R .* rₖ
+
+        # ∂²J/∂Δt² = (v - baseline)' * R * (v - baseline)
+        ∂²J[Δt_index, Δt_index] = dot(rₖ, reg.R .* rₖ)
+    end
+    
+    return ∂²J
 end
 
-function QuadraticRegularizer(
-    name::Symbol,
-    traj::NamedTrajectory,
-    R::Real;
-    kwargs...
-)
-    return QuadraticRegularizer(name, traj, R * ones(traj.dims[name]); kwargs...)
+# ============================================================================ #
+
+@testitem "testing QuadraticRegularizer" begin
+    include("../../test/test_utils.jl")
+    using DirectTrajOpt.Objectives
+
+    _, traj = bilinear_dynamics_and_trajectory()
+
+    R = 1.0
+    OBJ = QuadraticRegularizer(:u, traj, R)
+
+    test_objective(OBJ, traj, atol=1e-5)
```

Comment:
open seperate issue for regularizers and what we should test here, not combined with others